set timestamps to a fixed date(10+/-2)days ago for each cache_id

naiyoma opened the pull request:
Currently, there is a fingerprint attack via the responses to GETADDR messages. The spy can look at responses to GETADDR message from “different” nodes, match them up based on the timestamps returned, and see the size of overlap. The tor & clearnet nodes that have the most overlap are reliably tracing back to the same nodes. 

More details on this attack can be here  -> https://delvingbitcoin.org/t/fingerprinting-nodes-via-addr-requests/1786

This Pr mitigates this attack by setting the timestamps of different caches to a fixed value in the past(10+/-2) days  ensuring that there isn't a correlation using timestamps. 

Tested this using our own nodes 

before this change - overlap 80% after the change %.

This is still a work in progress as we are still testing and are open to trying to other solutions are well 
This is joint work with daniela

danielabrozzoni commented on src/net.cpp:
```diff
@@ -3552,6 +3552,13 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
         // nodes to be "terrible" (see IsTerrible()) if the timestamps are older than 30 days,
         // max. 24 hours of "penalty" due to cache shouldn't make any meaningful difference
         // in terms of the freshness of the response.
+
```
I think we should put here a long comment explaining why we're setting all the timestamps to the same fixed value, similar to the comment above explaining why it's useful to have a cache

danielabrozzoni reviewed (commented):
A few thoughts:

- In the PR description: "before this change - overlap 80% after the change %." -> I don't think it makes sense to say what is the overlap after the change, it is simply 0%, as the caches are returning completely different timestamps. Instead, What we can do is post an example of how the addr data would look like now, if we want people to better understand what's going on.

- I think it might be useful to add a couple of examples to the PR description. I have found that PRs that have a long description with examples and backgrounds are easier to review (for example: https://github.com/bitcoin/bitcoin/pull/32579)
I would expand the first paragraph so that people that aren't familiar with the addr cache can learn about it, it can be similar to the Delving Post background section. Then I would add an example section which can have a before and after: before, you show that in different caches the same address has the same timestamp, after, you show that each cache has its own timestamp and all the addresses have the same one.

- I think we can have a section in the PR description where we very briefly explain that we considered the 0 timestamp solution, but as far as we know it wasn't compatible with btcd. It can help motivate why we went for this solution.

- Finally, I would expand a functional test (maybe p2p_getaddr_caching?) to check that all the addresses in one network have the same timestamp.

naiyoma commented:
@danielabrozzoni , I updated and addressed your feedback above

With the current solution, different cache_ids sometimes produce the same random sequence, meaning the timestamps will be the same at some point. The probability of this happening is around 20% — when testing, it shows up at least once in every 5 getaddr requests.

Unfortunately, seeding FastRandomContext with the cache_id doesn’t work. But a seeded RNG would solve this, and I could also try InsecureRandomContext.

Could you test this and let me know what output you get?

danielabrozzoni commented:
Yeah, I get the same problem. One way to solve this is by shifting from a days granularity to a minutes granularity. Instead of just doing -10+-2 days, also manipulate the minutes.

So instead of having only these discrete timestamps (with a 1-day granularity):
13 days ago
12 days ago
11 days ago
10 days ago
9 days ago
8 days ago

You now have much more fine-grained timestamps, like:
13 days ago
12 days 23 hours 59 minutes ago
12 days 23 hours 58 minutes ago
12 days 23 hours 57 minutes ago
etc.

This way, you're not even leaking the cache creation time, as you were doing before.


The code would be:
`const auto new_timestamp = now - std::chrono::minutes(8 * 60 * 24 + FastRandomContext().randrange(5 * 60 * 24));`

From a very quick chatgpt query, it seems that the probability of having the same timestamp in two caches in this case is very low, about 0.01% :)
